### ISSUE TITLE
Config validation: Crash (segfault) when running in Docker

### ISSUE THREAD

**Issue by Firehed** [spiral/roadrunner#12]
I suspect this is just due to the base image being incompatible with the release artifact - having an official Docker image with a correctly-built binary may solve the problem (and would be far more convenient to use either way). I was also unsuccessful at building the binary from source in a Docker image using the provided instructions - seemed to be an issue with dependencies, but I'm not very familiar with building Go projects.

```Dockerfile
FROM php:7.2.6-alpine
RUN wget -O /tmp/rr.tar.gz https://github.com/spiral/roadrunner/releases/download/v1.0.1/roadrunner-1.0.1-linux-amd64.tar.gz \
    && tar -xzvf /tmp/rr.tar.gz -C /tmp \
    && rm -rf /tmp/rr.tar.gz \
    && cp /tmp/roadrunner-1.0.1-linux-amd64/rr /usr/local/bin/rr \
    && rm -rf /tmp/roadrunner-1.0.1-linux-amd
WORKDIR www
# for debugging
ENTRYPOINT sh
```

```sh
/www # rr --config .rr.yaml -v serve
DEBU[0000] [http]: started
DEBU[0000] [static]: started
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x800a2a]

goroutine 7 [running]:
github.com/spiral/roadrunner.(*Server).Start(0xc42009c960, 0x0, 0x0)
	/Users/wolfy-j/go/src/github.com/spiral/roadrunner/server.go:70 +0xea
github.com/spiral/roadrunner/service/http.(*Service).Serve(0xc42006a480, 0x0, 0x0)
	/Users/wolfy-j/go/src/github.com/spiral/roadrunner/service/http/service.go:88 +0x211
github.com/spiral/roadrunner/service.(*container).Serve.func1(0xc42008fb60, 0xc42006aba0, 0xc4201fe7e0)
	/Users/wolfy-j/go/src/github.com/spiral/roadrunner/service/container.go:142 +0x8e
created by github.com/spiral/roadrunner/service.(*container).Serve
	/Users/wolfy-j/go/src/github.com/spiral/roadrunner/service/container.go:138 +0x1de
```

**Comment by wolfy-j** [comment]
Ok, I managed to recreate this issue by corrupting config file. Can you make sure that http.workers.pool section presented? Can you post your config file sample here?

I have to add more validations for the config to track scenarios when config sections are omitted.

**Comment by Firehed** [comment]
Ah, that makes sense. I was providing the minimal config that didn't give a validation error, assuming it would fill in default values as necessary. Here's what I had:

```yaml
rpc:
  enable: false
http:
  enable: true
  address: 0.0.0.0:8000
  workers:
    command: php rr-entrypoint.php
    relay: pipes
  uploads:
    forbid: []
static:
  enable: true
  dir: public
  forbid:
    - .php
```

**Comment by wolfy-j** [comment]
That explains :) We have slipped couple validations but I will keep this ticket open until your config given a proper error. We do not have default values for pool options, they always are custom.

Try:

```yaml
rpc:
  enable: false
http:
  enable: true
  address: 0.0.0.0:8000
  workers:
    command: php rr-entrypoint.php
    relay: pipes
    pool:
       numWorkers: 4
       allocateTimeout: 600000000
       destroyTimeout:  600000000  
  uploads:
      forbid: []
static:
  enable: true
  dir: public
  forbid:
    - .php
```

**Comment by wolfy-j** [comment]
You can also write configs in json btw.

**Comment by Firehed** [comment]
Adding the `http.workers.pool` section appears to have fixed the crash-on-startup, but an immediately-dying worker (I made a typo and did `require_once 'vendor/autoloaad.php';`) just put this in the output, which is not terribly useful for debugging:

```
rr_1            | time="2018-06-19T19:15:56Z" level=error msg="[http]: unable to connect to worker: unexpected response, header is missing: signal: killed"
```

I was only able to figure out the actual source of the issue by trying to run the worker command myself in a shell; at no point was I able to actually find the PHP error itself when served by `rr`. Though to be fair, that could be due to bad `php.ini` settings - I haven't poked at it a whole lot.

Having additional documentation on the config would be super valuable here. For example, I have absolutely no idea how the RPC section works or what it's supposed to accomplish.

**Comment by wolfy-j** [comment]
That's weird, this has been properly tested: http://prntscr.com/jwvfyw

I would have to take a look at this docker container and build our own version later.

**Comment by wolfy-j** [comment]
@Firehed would you mind creating a ticket for that?
> Having additional documentation on the config would be super valuable here. For example, I have absolutely no idea how the RPC section works or what it's supposed to accomplish.

**Comment by Firehed** [comment]
Will do!

What are your log settings if you run `php -i`? The base image I started off with has fairly dumb defaults, which could be part of the issue.

If it's relevant, my application's internal logger goes to STDOUT (which is what Docker generally wants), and the front-controller script I'm trying `rr` with looks like this:

```php
// autoload, framework setup, etc...
$relay = new Spiral\Goridge\StreamRelay(STDIN, STDOUT);
$psr7 = new Spiral\RoadRunner\PSR7Client(new Spiral\RoadRunner\Worker($relay));

while ($req = $psr7->acceptRequest()) {
    try {
        $response = $dispatcher->setRequest($req)->dispatch();
        $psr7->respond($response);
    } catch (\Throwable $e) {
        $psr7->getWorker()->error((string)$e);
    }
}
```

**Comment by wolfy-j** [comment]
Try setting `ini_set('display_errors', 'stderr');` if your script, rr will handle error buffer in this case and forward it to parent docker container with proper error management before that (like displaying in the log).

In some cases (fatal errors on bootload stage) php dies even before stderr can be set, running script manually `php script.php` actually very solid solution to check if worker bootloads (script must be awaiting for stdin input).

**Comment by wolfy-j** [comment]
Config validation issue has been resolved in 1.0.2
